[MLIR][Linalg] Introduce broadcast/transpose semantic to 'linalg.batc…

shahidact opened the pull request:
…h_matmul' operation.

Goals:
1. To add syntax and semantic to 'batch_matmul' without changing any of the existing syntax expectations for current usage. batch_matmul is still just batch_matmul.

2. Move the definition of batch_matmul from linalg OpDsl to tablegen ODS infra.

Scope of this patch:
To expose broadcast and transpose semantics on the 'batch_matmul'.

The broadcast and transpose semantic are as follows:

By default, 'linalg.batch_matmul' behavior will remain as is. Broadcast and Transpose semantics can be applied by specifying the explicit attribute 'indexing_maps' as shown below. This is a list attribute, so the list must include all the maps if specified.

    Example Transpose:
    ```
    linalg.batch_matmul indexing_maps = [
                   affine_map< (d0, d1, d2, d3) -> (d0, d3, d1)>, //transpose
                   affine_map< (d0, d1, d2, d3) -> (d0, d3, d2)>,
                   affine_map< (d0, d1, d2, d3) -> (d0, d1, d2)>
                   ]
                   ins (%arg0, %arg1: memref<2x5x3xf32>,memref<2x5x7xf32>)
                   outs (%arg2: memref<2x3x7xf32>)
    ```

    Example Broadcast:
    ```
    linalg.batch_matmul indexing_maps = [
                       affine_map< (d0, d1, d2, d3) -> (d3)>,  //broadcast
                       affine_map< (d0, d1, d2, d3) -> (d0, d3, d2)>,
                       affine_map< (d0, d1, d2, d3) -> (d0, d1, d2)>
                     ]
                     ins (%arg0, %arg1: memref<5xf32>,memref<2x5x7xf32>)
                     outs (%arg2: memref<2x3x7xf32>)
    ```

    Example Broadcast and transpose:
    ```
    linalg.batch_matmul indexing_maps = [
                       affine_map< (d0, d1, d2, d3) -> (d1, d3)>,     //broadcast
                       affine_map< (d0, d1, d2, d3) -> (d0, d2, d3)>, //transpose
                       affine_map< (d0, d1, d2, d3) -> (d0, d1, d2)>
                     ]
                     ins (%arg0, %arg1: memref<3x5xf32>, memref<2x7x5xf32>)
                     outs (%arg2: memref<2x3x7xf32>)
    ```

RFCs and related PR:
https://discourse.llvm.org/t/rfc-linalg-opdsl-constant-list-attribute-definition/80149
https://discourse.llvm.org/t/rfc-op-explosion-in-linalg/82863
https://discourse.llvm.org/t/rfc-mlir-linalg-operation-tree/83586
https://github.com/llvm/llvm-project/pull/115319

shahidact committed:
[MLIR][Linalg] Introduce broadcast/transpose semantic to 'linalg.batch_matmul' operation.

Goals:
1. To add syntax and semantic to 'batch_matmul' without changing any of the
   existing syntax expectations for current usage. batch_matmul is still
   just batch_matmul.

2. Move the definition of batch_matmul from linalg OpDsl to tablegen ODS
   infra.

Scope of this patch:
To expose broadcast and transpose semantics on the 'batch_matmul'.

The broadcast and transpose semantic is as follows:

By default 'linalg.batch_matmul' behavior will remain as is.
Broadcast and Transpose semantics can be appiled by specifying the
explicit attribute 'indexing_maps' as shown below.This is a list attribute, so the list
must include all the maps if specified.

    Example Transpose:
    ```
    linalg.batch_matmul indexing_maps = [
                   affine_map<(d0, d1, d2, d3) -> (d0, d3, d1)>, //transpose
                   affine_map<(d0, d1, d2, d3) -> (d0, d3, d2)>,
                   affine_map<(d0, d1, d2, d3) -> (d0, d1, d2)>
                   ]
                   ins(%arg0, %arg1 : memref<2x5x3xf32>,memref<2x5x7xf32>)
                   outs(%arg2: memref<2x3x7xf32>)
    ```

    Example Broadcast:
    ```
    linalg.batch_matmul indexing_maps = [
                       affine_map<(d0, d1, d2, d3) -> (d3)>,     //broadcast
                       affine_map<(d0, d1, d2, d3) -> (d0, d3, d2)>,
                       affine_map<(d0, d1, d2, d3) -> (d0, d1, d2)>
                     ]
                     ins(%arg0, %arg1 : memref<5xf32>,memref<2x5x7xf32>)
                     outs(%arg2: memref<2x3x7xf32>)
    ```

    Example Broadcast and transpose:
    ```
    linalg.batch_matmul indexing_maps = [
                       affine_map<(d0, d1, d2, d3) -> (d1, d3)>,     //broadcast
                       affine_map<(d0, d1, d2, d3) -> (d0, d2, d3)>, //transpose
                       affine_map<(d0, d1, d2, d3) -> (d0, d1, d2)>
                     ]
                     ins(%arg0, %arg1 : memref<3x5xf32>, memref<2x7x5xf32>)
                     outs(%arg2: memref<2x3x7xf32>)
    ```

adam-smnk commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -3450,6 +3467,46 @@ static LogicalResult verifyExtendedMatmulSemantic(MatmulOp matmulOp,
+/// It checks if the first result dimension is a function of the first
+/// dimension.
+static bool isValidBatchDim(AffineMap bcastMap) {
+  assert(bcastMap.getNumResults() == 3 && "Expected three result dim expr.");
```
Is this guaranteed to pass or would it be safer to just return false?
Maybe add a test case that covers this.

shahidact replied:
The assert would be raised for this kind of LHS map "affine_map<(d0, d1, d2, d3) -> (d0, d1, d3, d3)>". I cannot return "False" for this case. The invariant here is that for batch matmul, number of result dims should not exceed 3.

shahidact replied:
Actually, in the context of the call site of this function, this assert is guaranteed to be True. Am I missing something here?

adam-smnk replied:
Using the example you provided trips the assertion:
```mlir
linalg.batch_matmul indexing_maps = [
                       affine_map<(d0, d1, d2, d3) -> (d0, d1, d3, d3)>,
                       affine_map<(d0, d1, d2, d3) -> (d0, d3, d2)>,
                       affine_map<(d0, d1, d2, d3) -> (d0, d1, d2)>
                      ]
                      ins(%arg0, %arg1 : memref<?x?x?xf32>, memref<?x?x?xf32>)
                      outs(%arg2 :memref<?x?x?xf32>)
```

shahidact replied:
Yes, it will. By definition of batch_matmul, user should not pass >3 result dim and hence thought of using assert. I got your point that you want this to be explicitly reported, will update.

shahidact replied:
Oh, missed to update.

adam-smnk commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -3450,6 +3467,46 @@ static LogicalResult verifyExtendedMatmulSemantic(MatmulOp matmulOp,
+  return exp.isFunctionOfDim(0);
+}
+
+/// Verifies the broadcast and transpose semantic sepecified by the explicit
```
```suggestion
/// Verifies the broadcast and transpose semantic specified by the explicit
```

adam-smnk commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -3611,5 +3668,165 @@ Speculation::Speculatability MatmulOp::getSpeculatability() {
+  if (!hasUserDefinedMaps())
+    return success();
+
+  for (unsigned opIndex = 0; opIndex < 2; opIndex++) {
```
Why not validate output maps as well?

Right now sth like this is possible:
```mlir
linalg.batch_matmul indexing_maps = [
                       affine_map<(d0, d1, d2, d3) -> (d0, d1, d3)>,
                       affine_map<(d0, d1, d2, d3) -> (d0, d3, d2)>,
                       affine_map<(d0, d1, d2, d3) -> (d1)>
                     ]
                     ins(%arg0, %arg1 : memref<5x5x5xf32>, memref<5x5x5xf32>)
                     outs(%arg2: memref<5xf32>)
```

Same goes for `linalg.matmul` with `indexing_maps`.

shahidact commented:
> Please, could you provide references to relevant RFCs where this was discussed/agreed-on? And include in the summary. Thanks :)

Added :)

shahidact committed:
-Added output map verification and corresponding tests.
-Replaced assert for the count of number of dim expression with proper error reporting and new test case.
-Fixed typos.

MaheshRavishankar commented:
Havent reviewed in detail yet. How is this related to https://discourse.llvm.org/t/mlir-rfc-introduce-linalg-contract/83589 .

rengolin commented:
> Havent reviewed in detail yet. How is this related to https://discourse.llvm.org/t/mlir-rfc-introduce-linalg-contract/83589 .

It is not. This is just a cleanup of OpDSL that we agreed last year. Please do not cross the wires here.

adam-smnk commented on mlir/lib/Dialect/Linalg/Transforms/DropUnitDims.cpp:
```diff
@@ -935,7 +935,8 @@ struct RankReduceContractionOps : OpRewritePattern<FromOpTy> {
     for (auto attr : contractionOp->getAttrs()) {
-      if (attr.getName() == LinalgDialect::kMemoizedIndexingMapsAttrName)
+      if (attr.getName() == LinalgDialect::kMemoizedIndexingMapsAttrName ||
+          attr.getName() == "indexing_maps")
```
I don't think it's safe to just drop user defined indexing maps.
Plus, could you add a relevant test case?

shahidact replied:
This change was to fix a crash in this existing test case "Dialect/Linalg/rank-reduce-contraction-ops.mlir".

adam-smnk replied:
Still, consider this example:
```mlir
func.func @entry(%arg0: memref<1x5x5xf32>,
    %arg1: memref<1x5x5xf32>, %arg2: memref<1x5x5xf32>) {
  linalg.batch_matmul indexing_maps = [
      affine_map<(d0, d1, d2, d3) -> (d0, d1, d3)>,
      affine_map<(d0, d1, d2, d3) -> (d0, d2, d3)>, // Transposed B
      affine_map<(d0, d1, d2, d3) -> (d0, d1, d2)>
    ]
    ins(%arg0, %arg1 : memref<1x5x5xf32>, memref<1x5x5xf32>)
    outs(%arg2: memref<1x5x5xf32>)
  return
}
```
Applying `mlir-opt -test-linalg-rank-reduce-contraction-ops` drops the maps which means transpose is lost.

shahidact replied:
I understand your concern; however, this pass is based on the default behaviour of the `batch_matmul`. In future, if I want to extend this pass for the extended behaviour I would handle it.

banach-space commented on mlir/test/Dialect/Linalg/generalize-named-ops.mlir:
```diff
@@ -1002,3 +1002,26 @@ func.func @matmul_transpose_a_b_explicit(%arg0: memref<5x3xf32>, %arg1: memref<7
+// CHECK:           arith.mulf
+// CHECK:           arith.addf
+
+func.func @batch_matmul(%arg0: tensor<?x?x?xf32>, %arg1: tensor<?x?x?xf32>, %arg2: tensor<?x?x?xf32>) -> tensor<?x?x?xf32> {
```
Looks like most/all examples in this file use static shapes - could you follow that convention?

Also, how about dedicated tests for transpose and broadcasts?

shahidact replied:
> Also, how about dedicated tests for transpose and broadcasts?

In the "matmul" PR I had got a comment that one test is enough to verify the generalization in conjunction with invalid tests, and it is true. Hence, I refrained to add redundancy.

banach-space commented:
> > Please, could you provide references to relevant RFCs where this was discussed/agreed-on? And include in the summary. Thanks :)
> 
> Added :)

Thanks! @MaheshRavishankar , relevant links have been added in the summary at the top:

> RFCs and related PR:
> https://discourse.llvm.org/t/rfc-linalg-opdsl-constant-list-attribute-definition/80149
> https://discourse.llvm.org/t/rfc-op-explosion-in-linalg/82863
> https://discourse.llvm.org/t/rfc-mlir-linalg-operation-tree/83586
> #115319

Looks like the syntax introduced here matches #115319 (unless I missed sth), so this is consistent with the previous patches in the series 👍🏻 

One high level (kind) request:

> Goals:
> 
> To add syntax and semantic to 'batch_matmul' without changing any of the existing syntax expectations for current usage. batch_matmul is still just batch_matmul.
> 
> Move the definition of batch_matmul from linalg OpDsl to tablegen ODS infra.

Why not split this into two patches? Those two goals seem orthogonal to me and reviewing small patches is easier (at least on GitHub).

rengolin commented:
> Why not split this into two patches? Those two goals seem orthogonal to me and reviewing small patches is easier (at least on GitHub).

From discussions on the previous PR (matmul), there should be no differences to what we did there. The ODS implementation is identical to the OpDSL one except the affine map. All previous tests should pass. So the extra logic that needs to be reviewed is actually a very small portion of the whole patch. If we split, the first patch would be almost as big as this one.

But getting the semantics correct is not trivial either. So splitting could lead to an interim state that is not quite here nor there, and we want to avoid that.

shahidact commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -3450,6 +3467,73 @@ static LogicalResult verifyExtendedMatmulSemantic(MatmulOp matmulOp,
+  SmallVector<AffineMap, 3> defaultIndexingMaps =
+      batchMatmulOp.getDefaultIndexingMaps(batchMatmulOp->getContext());
+
+  auto opIndexingMap = opIndexingMaps[opIndex];
```
If I undestand correctly, we do have such a test in "invalid.mlir" as below. However, let me know what case you have which crashes?

func.func @missing_indexing_map_batch_matmul(%arg0: memref<?x?x?xf32>, %arg1: memref<?x?x?xf32>, %arg2: memref<?x?x?xf32>) {
  // expected-error @+1 {{expected attribute value}}
  linalg.batch_matmul indexing_maps = [
                       ,
                       affine_map<(d0, d1, d2, d3) -> (d0, d3, d2)>,
                       affine_map<(d0, d1, d2, d3) -> (d0, d1, d2)>
                      ]
                      ins(%arg0, %arg1 : memref<?x?x?xf32>, memref<?x?x?xf32>)
                      outs(%arg2 :memref<?x?x?xf32>)
  return
}

adam-smnk replied:
You should test a few more possible variations.
For me, this one is problematic:
```mlir
linalg.batch_matmul indexing_maps = [
      affine_map<(d0, d1, d2, d3) -> (d0, d1, d3)>,
      affine_map<(d0, d1, d2, d3) -> (d0, d2, d3)>
    ]
    ins(%arg0, %arg1 : memref<5x5x1xf32>, memref<5x5x1xf32>)
    outs(%arg2: memref<5x5x5xf32>)
```

shahidact replied:
Sure

shahidact replied:
I will add tests.

MaheshRavishankar commented:
Thanks @shahidact looks good to me (i didnt review all the details, but mostly for structure of things and that matches what I expect)

shahidact committed:
*Added checks for extended semantics and exit gracefully in user passes.
*Added and udated test cases.
*Refactored verification logic.

shahidact commented:
> Thanks @shahidact looks good to me (i didnt review all the details, but mostly for structure of things and that matches what I expect)

Thanks a lot.

shahidact committed:
*Added logic and tests to verify the size of supplied indexing_map attribute.

adam-smnk commented on mlir/lib/Dialect/Linalg/Transforms/DropUnitDims.cpp:
```diff
@@ -906,6 +906,15 @@ struct RankReduceContractionOps : OpRewritePattern<FromOpTy> {
                                 PatternRewriter &rewriter) const override {
+    // Check to not let go the batch_matmul with extended semantic, through this
+    // transform.
+    if (std::is_same<FromOpTy, BatchMatmulOp>::value) {
```
Since we're already here, could you do the same for matmul?

shahidact replied:
Done

adam-smnk commented on mlir/lib/Dialect/Linalg/Transforms/DropUnitDims.cpp:
```diff
@@ -935,7 +944,8 @@ struct RankReduceContractionOps : OpRewritePattern<FromOpTy> {
     for (auto attr : contractionOp->getAttrs()) {
-      if (attr.getName() == LinalgDialect::kMemoizedIndexingMapsAttrName)
+      if (attr.getName() == LinalgDialect::kMemoizedIndexingMapsAttrName ||
+          attr.getName() == "indexing_maps")
```
Is this change still needed?
I think just dropping maps is a dangerous idea.

shahidact replied:
Yes, it is still needed. This pass, in this case is taking `batch_matmul `and transforming into `matmul` but not updating (probably due to default batch_matmul which did not have indexing_map attribute) the `indexing_map` which correspond to `batch_matmul`.  Hence the transformed matmul IR has indexing map of batch_matmul. I will fix it such that we don't have to drop this attribute entirely.

shahidact replied:
Done

banach-space commented on mlir/test/Dialect/Linalg/named-ops.mlir:
```diff
@@ -1487,6 +1487,154 @@ func.func @matmul_transpose_b(%arg0: memref<3x5xf32>, %arg1: memref<7x5xf32>, %a
+//       CHECK:   linalg.batch_matmul
+//  CHECK-SAME:     ins(%{{.+}}, %{{.+}} : memref<2x5x3xf32>, memref<2x5x7xf32>)
+//  CHECK-SAME:     outs(%{{.+}} : memref<2x3x7xf32>)
+func.func @batch_matmul_explicit_transpose_a(%arg0: memref<2x5x3xf32>, %arg1: memref<2x5x7xf32>, %arg2: memref<2x3x7xf32>) {
```
For consistency
```suggestion
func.func @batch_matmul_explicit_transpose_A(%arg0: memref<2x5x3xf32>, %arg1: memref<2x5x7xf32>, %arg2: memref<2x3x7xf32>) {
```

shahidact replied:
Done

banach-space commented on mlir/test/Dialect/Linalg/named-ops.mlir:
```diff
@@ -1487,6 +1487,154 @@ func.func @matmul_transpose_b(%arg0: memref<3x5xf32>, %arg1: memref<7x5xf32>, %a
+//       CHECK:   linalg.batch_matmul
+//  CHECK-SAME:     ins(%{{.+}}, %{{.+}} : memref<2x3x5xf32>, memref<2x7x5xf32>)
+//  CHECK-SAME:     outs(%{{.+}} : memref<2x3x7xf32>)
+func.func @batch_matmul_explicit_transpose_b(%arg0: memref<2x3x5xf32>, %arg1: memref<2x7x5xf32>, %arg2: memref<2x3x7xf32>) {
```
For consistency
```suggestion
func.func @batch_matmul_explicit_transpose_B(%arg0: memref<2x3x5xf32>, %arg1: memref<2x7x5xf32>, %arg2: memref<2x3x7xf32>) {
```

banach-space commented on mlir/test/Dialect/Linalg/named-ops.mlir:
```diff
@@ -1487,6 +1487,154 @@ func.func @matmul_transpose_b(%arg0: memref<3x5xf32>, %arg1: memref<7x5xf32>, %a
+// CHECK:           linalg.batch_matmul ins(%[[VAL_0]], %[[VAL_1]] : memref<5xf32>, memref<2x5x7xf32>) outs(%[[VAL_2]] : memref<2x3x7xf32>) indexing_maps = [#[[$ATTR_0]], #[[$ATTR_1]], #[[$ATTR_2]]]
+// CHECK:           return
+// CHECK:         }
+func.func @batch_matmul_bcast_batch_and_m_dim_A(%arg0: memref<5xf32>, %arg1: memref<2x5x7xf32>, %arg2: memref<2x3x7xf32>) {
```
[nit] This name suggests that the `batch` and `m` dims are broadcast:
* `@batch_matmul_bcast_batch_and_m_dim_A`. 

That's not quite true though, right? It's the `k` dim that's broadcast and `batch` and `m` are the missing dims.

shahidact replied:
IIUC, I think it otherwise. I intend `batch` and `m` gets broadcasted. So that matrix A varies as `d3`.

shahidact replied:
Oh, now I got what you meant :). I will rename it, what about this `batch_matmul_bcast_k_to_fill_missing_dims_in_A`

banach-space replied:
SGTM :)

shahidact replied:
Done.

banach-space commented on mlir/lib/Dialect/Linalg/Transforms/TransposeMatmul.cpp:
```diff
@@ -88,6 +88,14 @@ FailureOr<Operation *>
+  if (batchMatmulOp.hasUserDefinedMaps()) {
+    return rewriter.notifyMatchFailure(
+        batchMatmulOp,
+        "only batch_matmul ops with non-extended semantics are supported");
```
Either `linalg.batch_matmul` or skip altogether. IIRC, the Op will be printed anyway. Also:
```suggestion
        "user-defined maps are not supported");
```

shahidact replied:
Done

banach-space commented on mlir/lib/Dialect/Linalg/Transforms/TransposeMatmul.cpp:
```diff
@@ -88,6 +88,14 @@ FailureOr<Operation *>
 mlir::linalg::transposeBatchMatmul(RewriterBase &rewriter,
                                    linalg::BatchMatmulOp batchMatmulOp,
                                    bool transposeLHS) {
+  // Check to not let go the batch_matmul with extended semantic, through this
+  // transform.
```
[nit] Skip, this is clear from the code.

shahidact replied:
Done

banach-space commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td:
````diff
@@ -680,6 +680,130 @@ def MatmulOp : LinalgStructuredBase_Op<"matmul", [
+                     outs(%arg2: memref<2x3x7xf32>)
+    ```
+
+    Example Broadcast and transpose:
````
Please use capitalisation consistently
```suggestion
    Example Broadcast and Transpose:
```

shahidact replied:
Done

banach-space commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td:
```diff
@@ -680,6 +680,130 @@ def MatmulOp : LinalgStructuredBase_Op<"matmul", [
+  /*extraInterfaces=*/[LinalgContractionOpInterface])> {
+    
+  let summary = [{Performs a batched matrix multiplication of two 3D inputs.}];
+  let description = [{Numeric casting is performed on the operands to the inner multiply, promoting
```
IIUC, casting happens in `regionBuilder` and is hard-coded to "signed"?

shahidact replied:
Yes

shahidact replied:
I borrowed this comment from older definition. I think it intentionally did not document the details as this comment will be part of the dialect documentation.

banach-space replied:
This makes me sad, but I appreciate that you are trying to preserve as much of the original logic as possible. OK to ignore.

banach-space commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td:
```diff
@@ -680,6 +680,130 @@ def MatmulOp : LinalgStructuredBase_Op<"matmul", [
+    them to the same data type as the accumulator/output.
+
+    Broadcast and Transpose semantics can be appiled by specifying the explicit attribute
+    'indexing_maps' as shown below.This is a list attribute, so the list must include all
+    the maps if specified.
```
```suggestion
    'indexing_maps' as shown below. This is a list attribute, so must include
    maps for all arguments if specified.
```
Why do we need all maps? That's somehow intuitive, but I don't understand how it being implemented as "list attribute" is relevant here?

shahidact replied:
This is to be consistent with existing approach.

shahidact replied:
I am referring to the definition of `GenericOp`.

shahidact replied:
Done.

banach-space commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td:
````diff
@@ -680,6 +680,130 @@ def MatmulOp : LinalgStructuredBase_Op<"matmul", [
+    Example Broadcast:
+    ```
+    linalg.batch_matmul indexing_maps = [
+                       affine_map<(d0, d1, d2, d3) -> (d3)>,     // broadcast
+                       affine_map<(d0, d1, d2, d3) -> (d0, d3, d2)>,
+                       affine_map<(d0, d1, d2, d3) -> (d0, d1, d2)>
````
Please use consistent formatting for all examples.
```suggestion
                       affine_map<(d0, d1, d2, d3) -> (d3)>,              // broadcast
                       affine_map<(d0, d1, d2, d3) -> (d0, d3, d2)>,
                       affine_map<(d0, d1, d2, d3) -> (d0, d1, d2)>
```

shahidact replied:
Done

banach-space commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td:
```diff
@@ -680,6 +680,130 @@ def MatmulOp : LinalgStructuredBase_Op<"matmul", [
+        return regionBuilder;
+      }
+
+      /// Returns a list of AffineMap with the typical batch_matmul indexing charactristic.
```
```suggestion
      /// Returns a list with default AffineMap(s), i.e. without broadcasts and transpositions.
```

shahidact replied:
Done

banach-space commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td:
```diff
@@ -680,6 +680,130 @@ def MatmulOp : LinalgStructuredBase_Op<"matmul", [
+      static unsigned getNumRegionArgs();
+      bool hasDynamicIndexingMaps() { return true; }
+      std::string getLibraryCallName();
+      /// Check if the op has broadcast and/or transpose semantic. Returns true if the
+      /// user defined indexing maps are not equal to default map.
+      bool hasUserDefinedMaps();
```
The name of this hook is misleading, it should more like `hasDefaultMaps`. I've left similar comment here:
* https://github.com/llvm/llvm-project/pull/122753/files#r1917202562

Perhaps just leave a TODO to rename it?

shahidact replied:
This returns true if the explicit map is different from default map, i.e. if user has supplied a map which is not of default characteristic. So, I think this is proper. Am I missing something here?

shahidact replied:
Sure I will add. One thing to note is that the extended semantic requires to initialize the `indexing_map` with default map even if user does not supply any.

banach-space replied:
OK, thanks for clarifying! Now I see why I got confused. 
* I read `hasUserDefineMaps` as - has the user specified the maps? (no matter what the value).

However, this is more like:
* `hasUserDefinedMaps` - This Op has user-defined (i.e. non-default) maps.

How about:
```
// Returns `true` if user-defined maps (that are different to the default maps) are used. Returns `false` otherwise. Note, if the user defines maps that are identical to the default maps, this method returns `false`.
bool hasUserDefinedMaps();
```

shahidact replied:
Since this is an interface method, I would prefer to modify the description as suggested in the `LinalgInterfaces.td` and keep these comments as is without any TODO. Does it make sense?

banach-space replied:
Skipping TODO makes sense, but I'm not sure about `LinalgInterfaces.td`. My only concern is that the comment in `LinalgInterfaces.td` has to be generic and there's some additional context here that's also worth documenting.

Give it a try!

shahidact replied:
Done.

banach-space reviewed (commented):
Thanks for working on this, the overall structure is good!

I've skimmed through the easier part and left some minor comments. I'll try to go over the rest in the next day/two.

shahidact committed:
*Added logic to update the indexing_map attribute for collapsed MatmulOp.
*Updated test names and comments for consistency.

banach-space commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -3450,6 +3467,95 @@ static LogicalResult verifyExtendedMatmulSemantic(MatmulOp matmulOp,
+           << "no. of result dim expression cannot exceed 3.";
+
+  // Check if the requested broadcast is valid.
+  if (isBroadcasted(opIndexingMap, defaultIndexingMap)) {
```
I've checked [isBroadcasted](https://github.com/llvm/llvm-project/blob/3b001db4f9668cfa29572e5f1911ec7cef8b0ac2/mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp#L3420-L3424) - that hook merely compares the number of results. IIUC, it will return true for this case:

```cpp
      map1 = affine_map<(d0, d1, d2, d3) -> (d2, d3)>,
      map2 = affine_map<(d0, d1, d2, d3) -> (d2, d2, d3)>,
```

Is `map1` a broadcast of `map2`?

1. Is there a definition of what "broadcasting" means in the context of this file?
2. `isBroadcasting` sounds too broad and should be renamed.

shahidact replied:
> 1. Is there a definition of what "broadcasting" means in the context of this file?

Here broadcast indexing maps are defined in context of corresponding default indexing maps for the given Op. This way the check becomes very simple i.e just check the number of result dims. I think inline check would be better.

banach-space replied:
Thanks! This description would be very helpful somewhere in the code. 

Also, given the 2 conditions that you listed above, I think that you could introduce a hook, e.g. `isIndexingMapABroadcastOf(mapToCheck, mapToCompareAgainst)` that would check these two things. In fact, isn't that already covered by:
* `isValidResultDimExprs` + `isBroadcast`.?

So:
```cpp
bool isIndexingMapABroadcastOf(mapToCheck, mapToCompareAgainst) {
  return isValidResultDimExprs(mapToCheck, mapToCompareAgainst) && isBroadcast(mapToCheck, mapToCompareAgainst);
}
```

Something along these lines (though I would change names of `isValidResultDimExprs` and `isBroadcast`).

shahidact replied:
Done.

shahidact replied:
> But in this case, `explicitMap` is not a broadcast of `defaultMap`, right?
No. This is (defaultMap = affine_map<(d0, d1, d2, d3) -> (d2, d3)>) not a default map for this op.

Below is the default map for A, B and C matrices respectively.
indexing_maps = [
                            affine_map<(d0, d1, d2, d3) -> (d0, d1, d3)>,
                            affine_map<(d0, d1, d2, d3) -> (d0, d3, d2)>,
                            affine_map<(d0, d1, d2, d3) -> (d0, d1, d2)>
                           ]

shahidact replied:
> ```c++
> areResultExprsSubsetOf(explicitMap, defaultMap)
> ```

My intention with this method was to check that user is not supplying unexpected dim.

shahidact replied:
@banach-space `areResultExprsSubsetOf(explicitMap, defaultMap)` is not part of the broadcast check. Pls read it in isolation.

adam-smnk commented on mlir/lib/Dialect/Linalg/Transforms/DropUnitDims.cpp:
```diff
@@ -937,7 +947,19 @@ struct RankReduceContractionOps : OpRewritePattern<FromOpTy> {
         continue;
-      collapsedOp->setAttr(attr.getName(), attr.getValue());
+
+      // Update the indexing_maps attribute for the collapsed MatmulOp.
+      if (attr.getName() == "indexing_maps" &&
+          std::is_same<FromOpTy, BatchMatmulOp>::value &&
+          std::is_same<ToOpTy, MatmulOp>::value) {
+        SmallVector<Attribute, 3> indexingMapsAttr = llvm::map_to_vector(
+            MatmulOp::getDefaultIndexingMaps(rewriter.getContext()),
+            [](AffineMap map) -> Attribute { return AffineMapAttr::get(map); });
+        collapsedOp->setAttr(attr.getName(),
+                             rewriter.getArrayAttr(indexingMapsAttr));
```
Why is this special case needed at all when user defined maps are disallowed?
If the pass breaks without this extra snippet, I think it's just treating symptom instead of the root cause.

shahidact replied:
> Why is this special case needed at all when user defined maps are disallowed?
This is actually not the user defined maps case. Pls go through my explanation above about why it is failing.

adam-smnk replied:
Hmm, I see but that's really not scalable solution.

I think we could ensure that only default indexing maps are present for all ops (see the other comment) and then drop them as default maps should be equivalent to `LinalgDialect::kMemoizedIndexingMapsAttrName`.

shahidact commented on mlir/lib/Dialect/Linalg/Transforms/DropUnitDims.cpp:
```diff
@@ -906,6 +907,15 @@ struct RankReduceContractionOps : OpRewritePattern<FromOpTy> {
                                 PatternRewriter &rewriter) const override {
+    // Check to not let go the batch_matmul with extended semantic, through this
+    // transform.
+    if (std::is_same<FromOpTy, BatchMatmulOp>::value ||
+        std::is_same<FromOpTy, MatmulOp>::value) {
+      if (contractionOp.hasUserDefinedMaps()) {
```
Yes, it can for whichever Op implements it as it is an interface. By default it returns false.

banach-space commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -3450,6 +3467,95 @@ static LogicalResult verifyExtendedMatmulSemantic(MatmulOp matmulOp,
+  // Check the result dims are valid.
+  if (!isValidResultDimExprs(opIndexingMap, defaultIndexingMap))
+    return batchMatmulOp->emitOpError()
+           << "Unexpected dim expression in map result.";
```
Definitely more accurate, but I'd try something shorter:
* `Unexpected dim expression (outside the set of the default result dims)`.

But then it's not clear that this is for a specific indexing map (out of possible 3). So, perhaps:
* `Unexpected dim expression in one of the maps (outside the set of the default result dims for that map)`

But then that's quite long as well :(

shahidact replied:
Done

shahidact commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -3450,6 +3467,95 @@ static LogicalResult verifyExtendedMatmulSemantic(MatmulOp matmulOp,
+  return exp.isFunctionOfDim(0);
+}
+
+/// Checks if the given AffineMap's result dimensions are valid output result
+/// dimensions.
```
> 1. What does it mean to be "valid"?

For this Op, by definition, the valid output map is `affine_map<(d0, d1, d2, d3) -> (d0, d1, d2)>`

shahidact replied:
> IMO, it would be better to check these inline in `verifyOutputMap` where the context is still clear. This is not generic enough to deserve a hook.

Sure

shahidact replied:
Done.

banach-space commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -3450,6 +3467,95 @@ static LogicalResult verifyExtendedMatmulSemantic(MatmulOp matmulOp,
+  // Check if the requested broadcast is valid.
+  if (isBroadcasted(opIndexingMap, defaultIndexingMap)) {
+    if (!batchMatmulOp.isValidLhsRhsBroadcastMap(opIndexingMap, isLHS))
+      return batchMatmulOp->emitOpError() << "Invalid broadcast requested.";
```
My bad, missed it!

shahidact committed:
*Added logic to ensure the indexing_map attribute can be dropped for collapsed contraction op.
*Refactored some tests and methods for better naming, comments and readability.

adam-smnk commented on mlir/lib/Dialect/Linalg/Transforms/DropUnitDims.cpp:
```diff
@@ -906,6 +907,10 @@ struct RankReduceContractionOps : OpRewritePattern<FromOpTy> {
                                 PatternRewriter &rewriter) const override {
+    if (contractionOp.hasUserDefinedMaps()) {
+      return rewriter.notifyMatchFailure(
+          contractionOp, "ops with user-defined maps are not supported");
```
Please add new test cases that cover these checks.

adam-smnk reviewed (commented):
Thanks for the fixes 👍

LGTM % open conversations

shahidact committed:
*Added tests to check DropUnitDim transform is not being applied on contraction Op having user defined indexing_maps.

shahidact commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -3450,6 +3467,95 @@ static LogicalResult verifyExtendedMatmulSemantic(MatmulOp matmulOp,
+                                     AffineMap opIndexingMap,
+                                     AffineMap defaultIndexingMap, bool isLHS) {
+  // Check the result dims are valid.
+  if (!isValidResultDimExprs(opIndexingMap, defaultIndexingMap))
```
Done

shahidact commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -3450,6 +3467,95 @@ static LogicalResult verifyExtendedMatmulSemantic(MatmulOp matmulOp,
+  // Check for valid number of result dims of input maps.
+  if (opIndexingMap.getNumResults() > 3)
+    return batchMatmulOp->emitOpError()
+           << "no. of result dim expression cannot exceed 3.";
```
Done

shahidact commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -3450,6 +3467,95 @@ static LogicalResult verifyExtendedMatmulSemantic(MatmulOp matmulOp,
+static bool isValidBatchDim(AffineMap bcastMap) {
+  AffineExpr exp = bcastMap.getResult(0);
+  return exp.isFunctionOfDim(0);
+}
```
Done.

shahidact commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -3450,6 +3467,95 @@ static LogicalResult verifyExtendedMatmulSemantic(MatmulOp matmulOp,
+
+/// Checks if the given AffineMap's result dimensions are valid output result
+/// dimensions.
+static bool isValidOutputResultDim(AffineMap outputMap) {
```
Done.

shahidact commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -3450,6 +3467,95 @@ static LogicalResult verifyExtendedMatmulSemantic(MatmulOp matmulOp,
+    if (!batchMatmulOp.isValidLhsRhsBroadcastMap(opIndexingMap, isLHS))
+      return batchMatmulOp->emitOpError() << "Invalid broadcast requested.";
+  } else if (!isValidBatchDim(opIndexingMap)) {
+    return batchMatmulOp->emitOpError()
+           << "Invalid batch dimension expression.";
```
Tests exist as in `invalid_batch_dim_batch_matmul_a` in `invalid.mlir`

shahidact committed:
*Updated code comment related to broadcast map check.

adam-smnk commented on mlir/lib/Dialect/Linalg/Transforms/DropUnitDims.cpp:
```diff
@@ -32,6 +32,7 @@
 #include "llvm/ADT/SetVector.h"
 #include "llvm/Support/CommandLine.h"
 #include "llvm/Support/Debug.h"
+#include <type_traits>
```
nit: I think it's not needed anymore

shahidact replied:
Yes, removed thanks.

shahidact committed:
*Removed undesired header from DropUnitDims.cpp

banach-space reviewed (approved):
What I've reviewed so far LGTM % the remaining comments (please address before landing).  I’m relying on @adam-smnk for the rest 😅

I will be OOO for a week, so approving as is to unblock you. Please feel free to merge once all outstanding comments are resolved.

Last but not least, thank you for working on this, @shahidact - this is a non-trivial area, and your work will have a significant impact on Linalg 🙏🏻

shahidact commented on mlir/test/Dialect/Linalg/named-ops.mlir:
```diff
@@ -1487,6 +1487,154 @@ func.func @matmul_transpose_b(%arg0: memref<3x5xf32>, %arg1: memref<7x5xf32>, %a
+// CHECK:           linalg.batch_matmul ins(%[[VAL_0]], %[[VAL_1]] : memref<3x5xf32>, memref<2x5x7xf32>) outs(%[[VAL_2]] : memref<2x3x7xf32>) indexing_maps = [#[[$ATTR_0]], #[[$ATTR_1]], #[[$ATTR_2]]]
+// CHECK:           return
+// CHECK:         }
+func.func @batch_matmul_bcast_batch_dim_A(%arg0: memref<3x5xf32>, %arg1: memref<2x5x7xf32>, %arg2: memref<2x3x7xf32>) {
```
> Did you mean `@batch_matmul_bcast_m_to_fill_missing_batch_dim_A`?

I meant to broadcast row-column of matrix A across broadcast dim.

shahidact committed:
*Renames few variables and updates few comments

shahidact commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -3450,6 +3469,88 @@ static LogicalResult verifyExtendedMatmulSemantic(MatmulOp matmulOp,
+
+/// Verifies the broadcast and transpose semantic specified by the explicit
+/// indexing map for the BatchMatmulOp \p op for each operand specified by \p
+/// opIndex.
```
Done

shahidact commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -3435,11 +3451,14 @@ static LogicalResult verifyExtendedMatmulSemantic(MatmulOp matmulOp,
+  // Check if the user defined map is valid broadcast map. Here broadcast
+  // indexing maps are defined in context of corresponding default indexing maps
+  // for the given Op. This way the check becomes very simple i.e just check the
+  // number of result dims.
```
Done.
